daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala

```diff
@@ -556,10 +558,11 @@ object ScenarioLedger {
   * @param ledgerData              Cache for the ledger.
   */
 case class ScenarioLedger(
+    interfaceInstances: data.Relation[Identifier, Identifier],
```

Comment:
I'm not sure I like this, it doesn't really seem like a ledger concern to know about interface instances. I'm wondering if we could achieve the same goal by adding the interfaces in the `ScenarioRunner` instead, since the `compiledPackages` are already available there. That way we don't need to carry any new information in `ScenarioLedger`.

```diff
-     (vcoinst: VersionedContractInstance) => callback(vcoinst.unversioned), 
+     (vcoinst: VersionedContractInstance) => callback(addInterfaces(compiledPackages, vcoinst.unversioned)), 
```

https://github.com/digital-asset/daml/blob/e8c7e29ba238e573c77de306d2591b76305f61b4/daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/ScenarioRunner.scala#L428-L437

Reply:
>  it doesn't really seem like a ledger concern to know about interface instances

I am not completely sure what his you problem here.

Keep in mind they are two side of a Ledger, the part that response to the engine (which does not care of interface at all) and the ledger API (which cares). 

* First the ledger know about the packages (because it stores them). Taking into account that looking for type information inside a given package is legit for any application (we provide the `api-type-signature` reader to allow third party application to read such info), I do not see why a ledger could not look for those.

* To implement efficient indexing on the API, the ledger has to know about template and interface.

* Since retroactive instance can create new instance for contract already persisted, a ledger has to be aware of the new retroactive instance in every new package it upload.

* `PackageInfo#interfaceInstances` was implemented only for this purpose. You can check it is used only once in the `ledger` module.

Reply:
To be fair the design of `ScenarioRunner/ScenarioLedger/ScenarioLedgerApi` is a mess.

Reply:
> A real ledger knows about the packages, because it should store them. Taking into account that looking for type information inside a given package is legit for any application [...] I do not see why a ledger could not look for the interface instances mapping.

Thank you, this is a fair point. I understand better now.

> If you prefer I could pass the full set of packages and generate the mapping internally.

No, I don't think that would be an improvement.